src/LoRaWANNode.cpp

```diff
@@ -0,0 +1,552 @@
+    if(LoRa_SetAppID(key) != AT_OK) {
+      return 0;
+    }
+  }
+
+  if(appKey != NULL) {
+    keyCharToInt(appKey, key, 16);
+    if(LoRa_SetKey(AT_APPKEY, key) == AT_OK) {
+      ret = Lora_Join(OTAA_JOIN_MODE);
+      if(ret == AT_OK) {  // Device joined
+        return 1;
+      } else if(ret == AT_JOIN_SLEEP_TRANSITION) {
+        uint32_t timeout = millis();
+        do {
+          if((millis() - timeout) > DELAY_FOR_JOIN_STATUS_REQ) {
+            return 0;
+          }
+          ret = Lora_JoinAccept();
+        } while(ret != AT_OK);
+        return 1;
+      }
+    }
+  }
+  return 0;
+}
+
+/*
+ * @brief  Send a LoRaWAN frame to the network.
+ * @param  pointer to the frame to send. Size is limited to 64 bytes.
+ * @param  Number of data to send.
+ * @param  If true the module requires a acknowledgment from the network.
+ * @param  The application port number.
+ * @retval LoRa error code.
+ */
+int8_t LoRaWANNodeClass::sendFrame(char frame[], uint8_t length, bool confirmed, uint8_t port)
+{
+  if((frame == NULL) || (length == 0)) {
+    return LORA_SEND_ERROR;
+  }
+
+  //Payload length is limited to 64 bytes.
+  if(length > MAX_PAYLOAD_LENGTH) {
+    return LORA_SEND_ERROR;
+  }
+
+  sSendDataBinary_t dataString = {frame, length, port, confirmed};
+  ATEerror_t errorCode = Lora_SendDataBin(&dataString);
+  if(errorCode == AT_OK) {
+    return length;
+  } else if(errorCode == AT_ERROR_WAN_SEND) {
+    // Previous frame not sent yet (busy or duty cycle). Wait before to try again.
+    return LORA_SEND_DELAYED;
+  }
+  return LORA_SEND_ERROR;
+}
+
+/*
+ * @brief  Receives a LoRaWAN frame from the network. Must be called before the
+ *         next call of sendFrame(). The receive delay depends on the Rx1 and Rx2
+ *         windows.
+ * @param  pointer to the received frame.
+ * @param  Number of data received.
+ * @param  The application port number.
+ * @retval Error code: false if failed else true.
+ */
+bool LoRaWANNodeClass::receivedFrame(uint8_t *frame, uint8_t *length, uint8_t *port)
+{
+  uint8_t data[128];
+
+  if((frame == NULL) || (length == NULL) || (port == NULL)) {
+    return 0;
+  }
+
+  sReceivedDataBinary_t structData = {data, 0, 0};
+  *length = 0;
+  if(Lora_AsyncDownLinkData(&structData) == AT_OK) {
+    if(structData.DataSize > 0) {
+      *length = structData.DataSize;
+      *port = structData.Port;
+      keyCharToInt((char *)data, frame, *length);
+      return 1;
+    }
+  }
+  return 0;
+}
+
+/*
+ * @brief  Read the LoRaWAN stack version.
+ * @param  pointer to version number (an Arduino string).
+ * @retval None.
+ */
+void LoRaWANNodeClass::getVersion(String *str)
+{
+  if(str != NULL) {
+    char tmp[10] = {'\0'};
+    Lora_GetVersion((uint8_t *)tmp);
+    str->concat(tmp);
+  }
+}
+
+/* NOTE: verbose mode is enabled when the 'ATI' command is sent. But it is
+  impossible to remove it (even if +VERB=0 is sent??!!). The verbose mode must be
+  disable if not the AT_VSSCANF function doesn't work.
```

Comment:
Verbose mode can't be used because the AT command driver provides by ST expects a specific format because the AT_VSSCANF function is a tiny implementation and is not able to parse some characters. Verbose mode must be off.

Reply:
OK. Please complete the existing comment with more details